Fix packed LUID conversions

mattst88 opened the pull request:
Two commits. I'm on Linux, so I'm unable to test the Windows-only changes. Hopefully CI will handle that.

---
Add and use packed_luid() function

Nearly all of the open-coded implementations were wrong in multiple ways:

- shifting by 31 instead of 32
- shifting a 32-bit value (the HighPart) without first casting to a 64-bit type
- shifting into the sign bit (undefined behavior)

---
Use packed_luid() to avoid aliasing violation

This solves the compile error when building with `-Werror=strict-aliasing`.

```
/home/mattst88/projects/gfxreconstruct/framework/encode/parameter_encoder.h: In member function ‘void gfxrecon::encode::ParameterEncoder::EncodeLUIDValue(LUID)’:
/home/mattst88/projects/gfxreconstruct/framework/encode/parameter_encoder.h:80:53: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
   80 |     void EncodeLUIDValue(LUID value) { EncodeValue(*reinterpret_cast<int64_t*>(&value)); }
      |                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Closes: https://github.com/LunarG/gfxreconstruct/issues/2358

---

cc: @igorbrsn, @JerryAMD, @rurra-amd because they all introduced instances of `<< 31`
cc: @jzulauf-lunarg because he introduced the aliasing violation and then dutifully ignored issue #2358

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/LunarG/gfxreconstruct?pullRequest=2559) <br/>All committers have signed the CLA.

ci-tester-lunarg commented:
Author mattst88 not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
Author mattst88 not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
CI gfxreconstruct build queued with queue ID 601414.

ci-tester-lunarg commented:
CI gfxreconstruct build # 8350 running.

lunarpapillo commented:
Fails to build on Wndows.  A typical `cmake` invocation is attached.

[006-cmake-stdout.txt](https://github.com/user-attachments/files/24159389/006-cmake-stdout.txt)

ci-tester-lunarg commented:
CI gfxreconstruct build # 8350 failed.

bradgrantham-lunarg commented:
> Fails to build on Wndows. A typical `cmake` invocation is attached.
> 
> [006-cmake-stdout.txt](https://github.com/user-attachments/files/24159389/006-cmake-stdout.txt)

@lunarpapillo The GitHub CI build also builds on Windows and I've enabled several times and the author has been able to provide several rounds of attempts to pass CI through branch pushes.  There's no need to tie up internal Jenkins CI results until the GH CI runs succeed.

mattst88 commented:
Could use another CI workflow approval. Hopefully this is the last time ;)

ci-tester-lunarg commented:
CI gfxreconstruct build # 8369 failed.

mattst88 commented:
@bradgrantham-lunarg: ping²?

ci-tester-lunarg commented:
Author mattst88 not on autobuild list. Waiting for curator authorization before starting CI build.

locke-lunarg commented:
@mattst88  Sorry. I have to revert your PR for now. Could you please restore your code for Windows? I will fix the issue for Windows.

mattst88 commented:
If the reason for this revert is that you also want the other change that I removed from this PR... then that makes no sense.

The change I removed my that PR is in this branch, if you want to take it over: https://github.com/mattst88/gfxreconstruct/tree/fix-luid-conversions

mattst88 commented:
By the way, if you're concerned about your CI costs, you should investigate why your macOS builds are taking 3 hours. At $0.062/minute, you're looking at $22 per CI run for just macOS, with everything else coming in at about $0.50 combined.
